test/common/bidi.py

```diff
@@ -0,0 +1,438 @@
+        self.download_dir = Path(self.homedir.name) / 'Downloads'
+        self.download_dir.mkdir()
+
+    async def start_bidi_session(self) -> None:
+        raise NotImplementedError('must be implemented by concrete subclass')
+
+    async def close_bidi_session(self) -> None:
+        raise NotImplementedError('must be implemented by concrete subclass')
+
+    async def close(self) -> None:
+        assert self.bidi_session is not None
+        log_proto.debug("cleaning up webdriver")
+
+        self.task_reader.cancel()
+        del self.task_reader
+        await self.ws.close()
+        await self.close_bidi_session()
+        self.bidi_session.process.terminate()
+        await self.bidi_session.process.wait()
+        self.bidi_session = None
+        await self.http_session.close()
+
+    def ws_done_callback(self, future):
+        for fut in self.pending_commands.values():
+            if not fut.done():
+                fut.set_exception(WebdriverError("websocket closed"))
+        if not future.cancelled():
+            log_proto.error("ws_reader crashed: %r", future.result())
+
+    async def start_session(self) -> None:
+        self.http_session = aiohttp.ClientSession(raise_for_status=True)
+        await self.start_bidi_session()
+        assert self.bidi_session
+        self.ws = await self.http_session.ws_connect(self.bidi_session.ws_url)
+        self.task_reader = asyncio.create_task(self.ws_reader(self.ws), name="bidi_reader")
+        self.task_reader.add_done_callback(self.ws_done_callback)
+
+        await self.bidi("session.subscribe", events=[
+            "log.entryAdded", "browsingContext.domContentLoaded",
+        ])
+
+        # wait for browser to initialize default context
+        for _ in range(10):
+            realms = (await self.bidi("script.getRealms"))["realms"]
+            if len(realms) > 0:
+                self.top_context = realms[0]["context"]
+                self.context = self.top_context
+                break
+            await asyncio.sleep(0.5)
+        else:
+            raise WebdriverError("timed out waiting for default realm")
+
+    async def __aenter__(self):
+        await self.start_session()
+        return self
+
+    async def __aexit__(self, *_excinfo):
+        if self.bidi_session is not None:
+            await self.close()
+
+    async def ws_reader(self, ws: aiohttp.client.ClientWebSocketResponse) -> None:
+        async for msg in ws:
+            if msg.type == aiohttp.WSMsgType.TEXT:
+                data = json.loads(msg.data)
+                log_proto.debug("ws TEXT → %r", data)
+                if "id" in data and data["id"] in self.pending_commands:
+                    log_proto.debug("ws_reader: resolving pending command %i", data["id"])
+                    if data["type"] == "success":
+                        if data["result"].get("type") == "exception":
+                            self.pending_commands[data["id"]].set_exception(
```

Comment:
Ditto.  I'd do something like `pending_command = self.pending_commands.get(tag)`.  It'd difficult to read this code because of the repetition of this very wide block of text (`...self.pending_commands[data["id"]]...`)

Reply:
OK, introduced `pending_command` alias.